File: examples/python/micro/gateway.py
from flask import request
from .user_service import get_user, update_user


def audited(func):
    return func


@audited
def handle_request():
    token = request.args.get("token")  # source: user input
    action = request.args.get("action")  # source: user input

    user = get_user(token)  # flows to SQL injection in auth_service
    result = update_user(token, action)  # flows to command injection in auth_service

    return {"user": user, "result": result}

File: examples/csharp/micro/Gateway.cs
using Microsoft.AspNetCore.Http;
using System.Collections.Generic;

namespace Micro
{
    public class AuditedAttribute : System.Attribute { }

    public class Gateway
    {
        private readonly UserService _userService = new UserService();

        [Audited]
        public string HandleRequest(HttpRequest req)
        {
            var token = req.Query["token"].ToString();    // source: user input
            var action = req.Query["action"].ToString();  // source: user input

            var user = _userService.GetUser(token);       // flows to SQL injection
            var result = _userService.UpdateUser(token, action); // flows to command injection

            return $"{{\"user\": \"{user[\"id\"]}\", \"result\": \"{result}\"}}";
        }
    }
}
